src/store/actions/Chat.ts

```diff
@@ -0,0 +1,161 @@
+export const SEND_ROOM_CHAT: string = "SEND_ROOM_CHAT";
+export const SEND_PRIVATE_CHAT: string = "SEND_PRIVATE_CHAT";
+export const SEND_TABLE_CHAT: string = "SEND_TABLE_CHAT";
+
+interface BaseSendChatFields {
+  venueId: string;
+  text: string;
+  from: string;
+}
+interface BaseSendChatToFields extends BaseSendChatFields {
+  to: string;
+}
+
+interface SendGlobalChatFields extends BaseSendChatFields {}
+interface SendPrivateChatFields extends Omit<BaseSendChatToFields, "venueId"> {}
+interface SendRoomChatFields extends BaseSendChatToFields {}
+interface SendTableChatFields extends BaseSendChatToFields {}
+
+interface SendGlobalChatAction extends SendGlobalChatFields {
+  type: typeof SEND_GLOBAL_CHAT;
+}
+interface SendPrivateChatAction extends SendPrivateChatFields {
+  type: typeof SEND_PRIVATE_CHAT;
+}
+interface SendRoomChatAction extends SendRoomChatFields {
+  type: typeof SEND_ROOM_CHAT;
+}
+interface SendTableChatAction extends SendTableChatFields {
+  type: typeof SEND_TABLE_CHAT;
+}
+
+enum ChatMessageType {
+  Global = "global",
+  Room = "room",
+  Private = "private",
+  Table = "table",
+}
+
+type BuildMessageInput = {
+  type: ChatMessageType.Global;
+  text: string;
+  from: string;
+};
+
+type BuildMessageToInput = {
+  type: ChatMessageType.Private | ChatMessageType.Room | ChatMessageType.Table;
+  text: string;
+  from: string;
+  to: string;
+};
+
+export type BaseChatMessage = {
+  from: string;
+  text: string;
+  ts_utc: firebase.firestore.Timestamp;
+  deleted?: boolean;
+};
+
+type NonGlobalChatMessage = BaseChatMessage & {
+  to: string;
+};
+
+export type GlobalChatMessage = BaseChatMessage & {
+  type: ChatMessageType.Global;
+};
+
+export type RestrictedChatMessage = NonGlobalChatMessage & {
+  type: ChatMessageType.Table | ChatMessageType.Room;
+};
+
+export type PrivateChatMessage = NonGlobalChatMessage & {
+  type: ChatMessageType.Private;
+  isRead: boolean;
+};
+
+export type ChatMessage =
+  | GlobalChatMessage
+  | PrivateChatMessage
+  | RestrictedChatMessage;
+
+const collectionName = (venueId: string) => `venues/${venueId}/chats`;
+
+const buildMessage = (
+  data: BuildMessageInput | BuildMessageToInput
+): ChatMessage => {
+  const base: BaseChatMessage = {
+    ts_utc: firebase.firestore.Timestamp.fromDate(new Date()),
+    text: data.text,
+    from: data.from,
+  };
+
+  switch (data.type) {
+    case ChatMessageType.Global:
+      return { ...base, type: data.type };
+    case ChatMessageType.Room:
+    case ChatMessageType.Table:
+      return { ...base, type: data.type, to: data.to };
+    case ChatMessageType.Private:
+      return { ...base, type: data.type, to: data.to, isRead: false };
+  }
+};
+
+const saveMessage = async (venueId: string, message: ChatMessage) =>
+  await firebase
+    .firestore()
+    .collection(collectionName(venueId))
+    .add(buildMessage(message));
+
+export const sendGlobalChat = createAsyncThunk<void, SendGlobalChatFields>(
```

Comment:
Up for a discussion:

Is it possible that sendMsgToGlobalChat is more descriptive?

Reply:
Good thinking here. It's more descriptive, yes; and it might help a reader out a little. But I suspect that since the meaning is the same, it makes sense to go with the more minimal expression:

* `sendGlobalChat` can be read as "send a global chat", where "chat" is the object of the sentence.
* `sendMsgToGlobalChat` can be read as "send a message to the global chat", where "message" is the object of the sentence.

They are both identical in meaning, but the shorter one is no more ambiguous, and I'm all for less code to read.

This is a really good suggestion, thanks for stimulating the discussion, very grateful you're suggesting these improvements.

My leaning here is toward the shorter `sendGlobalChat`. But happy to be convinced otherwise.

Reply:
Agreed